[IV-25-26] Objetivo 4

jorgelopez-ugr opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

jorgelopez-ugr committed:
Fix - Entrega mínima objetivo 4

Se añaden los ficheros necesarios para pasar los test

jorgelopez-ugr committed:
fix para pasar test

deno test missing

jorgelopez-ugr committed:
Merge branch 'main' into Objetivo-4

jorgelopez-ugr committed:
Fix a issue sin cerrar

Closes #28

jorgelopez-ugr committed:
ejemplo de codigo para scrapear

Se proporciona un ejemplo que contiene los datos en html para que pueda usarse como referencia para crear el scraper. Ha sido obtenido de la web de infaoliva. Closes #38

jorgelopez-ugr committed:
Queda documentada la elección de herramientas

En el texto aportado quedan descritas las herramientas que deben seleccionarse con sus respectivas definiciones. Se toma una decisión para cada una justificada por el único requisito definido, la integración nativa con el runtime deno. Closes #40

jorgelopez-ugr committed:
Se actualiza el readme con la referencia a las justificaciones de test

Por lo descrito en la #31 es necesario confirmar en el main la existencia de un archivo de justificaciones y comentar donde puede encontrarse tal archivo. Closes #31

jorgelopez-ugr committed:
Creación de funciones para clasificar los errores principales

Hay una serie de errores que se espera que sean disparados por mis funciones. Estos son, error de formato, error de tipo, error por valor faltante. Estos errores serán disparados cuando ocurra una de estas situaciones muy comunes cuando se trabaja con test. Los test que sean implementados deberán llevar al límite las funciones y eso implica pasarles valores erroneos faltantes y de todo tipo. Los test buscarán capturar y confirmar que estos errores estan siendo lanzados. Al crear un archivo solo de errores nos permite aislar toda la implementación en un solo sitio ayudandonos a desarrollarla y mejorando la modularidad del código. Sirve de ayuda para las issues: #47 #46 #48

jorgelopez-ugr committed:
Se añade la clase del scraper, el input de ejemplo, y las primeras funciones

En este commit se presentan los primeros datos de la fuente funcionando estos como ejemplo para comenzar el desarrollo. Se presenta la primera función auxiliar para limpiar etiquetas html llamada limpiarHTML. Consiste en una funcion que elimina la etiqueta y deja el campo realmente importante limpio de caracteres propios del html. Se presenta la primera función de extración, extraerFechaDeFila. Pertenece a la #47 pero esta se cerrará con la implementación y comprobación de los test correspondientes a estas dos funciones

jorgelopez-ugr committed:
funcion para extraer la variedad de una columna

Se define una funcion que va a servir para extraer las variedades de un html que se le pase. Forma parte, al igual que las anteriores, de la funcion del scraper que va a extraer los datos de la fuente. Pertenece a la #46. La funcion contiene todo tipo de comprobaciones pero se necesitan test para considerarla correcta.

jorgelopez-ugr committed:
funcion para extraer precios de la columna

funcion que se le pasa un codigo html que previamente ha sido determiando que se trata de una columna de precio, lo limpia de etiquetas y extrae el precio. Pertenece a la #48

jorgelopez-ugr committed:
funcion que parsea el precio y lo deja como dato usable

se ha definido esta funcion para dar solucion a lo especificado en la #49. Tenemos ahi un dato con el que se deben hacer calculos. Para hacerlos necesitamos que ese dato sea util, no nos vale una cadena de caracteres. Esta funcion concentra toda la funcionalidad correspondiente al parseo del precio. Elimina espacios y el simbolo del euro, lo convierte a numerico y lo devuelve. Contiene ademas todo tipo de comprobaciones. Deberán por supuesto crearse test para esta funcionalidad.

jorgelopez-ugr committed:
Definida una funcion para detectar cuando estamos ante una fila de fecha

esta funcion pertenece a la #50 y se trata de un codigo capaz de detectar cuando una fila que esta siendo parseada es en si una fila de fecha y por lo tanto representa el valioso codigo del que extraemos todos los datos. Indica el inicio de un bloque que podemos separar en 3 columnas principales. Para detectarlo busca la class=table-secondary presente en todas las filas que nombramos como fila fecha

jorgelopez-ugr committed:
funcion para extraer las columnas

La fuente de datos tiene forma de tabla, se propone esta funcion para extraer estas columnas conteniendo cada una un tipo de dato relevante. Pertenece a la #45 y será válida cuando los test correspondientes lo demuestren

jorgelopez-ugr committed:
Se extraen datos de la fuente con una funcion que reune las anteriores

LA funcion myOwnParser va a combinar todas las funcionalidades anteriores para que, con la menor complejidad posible, consiga extraer los datos de la fuente y almacenarlos en los objetos del Objetivo 2. Se trata de la representación combinada de toda la lógica de negocio y sería una simulación de lo necesario para dar valor al usuario de la #2. Pertenece a la #37 y será completado cuando se implemente y compruebe el test correspondiente.

jorgelopez-ugr committed:
test para la funcion limpiarHTML

Pertenece a la #45, funciona como una funcion auxiliar para limpiar texto de etiquetas html. Pasa todos los test.

jorgelopez-ugr committed:
test para la funcion extraerFechaDeFila

Estos son los test para la funcion que extrae las fechas de las filas. Se ha probado variedad de casos diferentes. Pasamos por el optimo, un caso en el que no hay ninguna fecha, uno en el que se le pasa ruido, uno vacio... Con todos estos test podemos verificar que la funcion hace lo que le pedimos y es robusta. Closes #47

jorgelopez-ugr committed:
test para probar la funcion extraerVariedadDeColumna

Una vez mas, probamos la funcion con casos extremos. La funcion lo supera todo correctamente. Closes #46

jorgelopez-ugr committed:
tests para la funcion extraerPrecioColumna

Se le pasan casos extremos incluso vacio. Si la funcion pasa estos test se puede considerar correcta y robusta, los pasa sin problema. Closes #48

jorgelopez-ugr committed:
test para la funcion parsearPrecio

esta funcion describe una problematica surgida durante el desarrollo. Se han creado una serie de tests para validarla que tienen en cuenta casos extremos. Los test son pasados sin problema por mi implementacion. Closes #49

jorgelopez-ugr committed:
test para la funcion esFilaFecha

test para la funcion que detecta si una fila es fila fecha. Valora multutud de casos diferentes. Son pasados sin problema Closes #50

jorgelopez-ugr committed:
test para la funcion que obtiene columnas de una fila

Es una de las funciones principales. Se le prueban casos problematicos como el de una entrada vacia completa o vacia parcial. Si los pasa todos la funcion esta probada y es robusta. Los pasa sin problema Closes #45

jorgelopez-ugr committed:
tests para la funcion principal myOwnParser

Se trata de una serie de test incluso mas minuciosos que los anteriores que prueban a fondo las distintas partes de la funcion principal. Verifican que el objeto devuelto existe y es correcto. Todos juntos componen el deseo del usuario de la #2. Como los pasa todos sin problema consideramos que Closes #37 dado que extrae los datos de la fuente y los convierte en los objetos válidos del objetivo 2.

jorgelopez-ugr committed:
todo confirmado en el readme

La guia de utilizacion esta finalmente presente en el README.md Closes #36

jorgelopez-ugr committed:
cambios reflejados en iv.yaml

Los cambios se reflejan tambien en iv.yaml

jorgelopez-ugr commented:
@gabrielherreraloz @gosema esto esta finalmente listo para revision. Comentadme que tal os parece

quedo atento al feedback

gabrielherreraloz reviewed (commented):
El código en sí lo veo bastante bien, y muy bien comentado, pero lo veo muy largo, te comento lo que me han comentado a mí y he visto en general, intenta reducir la lógica aquí, no hace falta que realices tantas tareas del tirón, recuerda que es un PMV, así que intenta quitar cosas, como extraer solo cierta información en vez de todo directamente. El código está correcto así que tecnicamente no lo pondría como un error, ¿que opinas de esto @jj? .

En cuanto a los commits, veo que mencionas los issues asociados en las descripciones pero en los títulos no los estás mencionando, eso me parece importante hacerlo para que sea todo más visual y sencillo de comprender.

jorgelopez-ugr commented:
@JJ te voy etiquetando ya porque el tiempo apremia. 

Muchas gracias.

JJ commented:
> El código en sí lo veo bastante bien, y muy bien comentado, pero lo veo muy largo, te comento lo que me han comentado a mí y he visto en general, intenta reducir la lógica aquí, no hace falta que realices tantas tareas del tirón, recuerda que es un PMV, así que intenta quitar cosas, como extraer solo cierta información en vez de todo directamente. El código está correcto así que tecnicamente no lo pondría como un error, ¿que opinas de esto @JJ? .

Ya lo dije en clase, sólo un PMV bien hecho. De hecho mirad el de @gosema, que en realidad ha extraido el semestre (aunque tampoco era fácil)

JJ commented on docs/tests.md:
```diff
@@ -0,0 +1,46 @@
+
+Para el desarrollo del objetivo se precisa decidir las sigueintes herramientas de tests:
+
+1. **Aserciones:** Son funciones que comparan la salida obtenida con la deseada, añadiendo un mensaje personalizado si la comparación es positiva. Las aserciones son simplemente funciones que validan condiciones específicas dentro de los tests.
```
Tío, no me cuentes lo que son. Muéstrame que tienes criterios para elegir herramientas que hagan esto. Esto *nunca* lo he considerado en esta asignatura, y lo sabéis.

JJ commented on docs/tests.md:
```diff
@@ -0,0 +1,46 @@
+## Análisis de opciones:
+
+# 1. Biblioteca de aserciones:
+- **Deno Assertions**
```
Podías poner el nombre de la biblioteca que se usa para importar

JJ commented on lib/domain/Error.js:
```diff
@@ -0,0 +1,16 @@
+// Archivo para contemplar los errores que van a devolverse en la terminal
+// en forma de funciones auxiliares para el Scraper.js
+
+function errorParsing(message) {
```
Las funciones no son clases, ni devuelve objetos que puedan interpretarse correctamente. Los errores forman parte del análisis del dominio.

JJ commented on lib/domain/Precios.js:
```diff
@@ -14,15 +14,62 @@ class VariedadPrecio {
         this.listaPrecios = new Map();
     }
+
+    // constructor(listaPrecios) {
```
¿Qué buscáis poniendo código comentado? NO lo entiendo. Es añadir ruido.

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+
+// --- Lógica del Parser ---
+
+const htmlInputSample = `
```
¿Esto lo pones en el código? ¿Un ejemplo? ¿Te parece eso una buena práctica?

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+        throw missingValue("Input no puede estar vacío");
+    }
+    
+    // Caso problematico: asd<strdasdong>asjdkh bakj ahs</stdsdrong>asd
```
Un caso problemático es literalmente un problema. ¿Y dónde se especifican los problemas?

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+ * @param {string} rawString 
+ * @returns {string}
+ */
+function limpiarHTML(rawString) {
```
¿Te parece que con todos los ifs que hay esto tenga una sola responsabilidad?

jorgelopez-ugr replied:
Sinceramente no se como ha llegado esa función ahí ya que la eliminé al hacer la simplificación. Cuando haga el proximo push me aseguraré mejor de que no quede nada que ya no sea valido.

JJ replied:
Se me olvidó subir los comentarios y cuando lo hice ya lo habías cambiado...

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+    rawString = rawString.trim();
+    
+    // Validación adicional: si después de limpiar queda vacío, lanzar error
+    if (rawString === '') {
```
¿Te parece una buena práctica usar siempre la misma variable para esto? En general, ¿te parece una buena práctica trabajar sobre el argumento de entrada pasado por referencia?

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+        throw missingValue("Input no puede estar vacío");
+    }
+
+    return rawString;
```
Es decir, no solo lo cambias sino que lo devuelves...

jorgelopez-ugr replied:
Mal hecho por mi parte se me habia pasado por completo. Esa funcion ya habia sido eliminada. Me aseguraré para la siguiente.

JJ commented on lib/domain/Scraper.js:
```diff
@@ -0,0 +1,238 @@
+        return match[1]; 
+    }
+    
+    throw errorParsing("Formato no correcto o no fecha valida en toda la fila");
```
Esto llama a una función que genera un error no específico. ¿Te parece que es una buena modelización del dominio?

jorgelopez-ugr committed:
Simplificacion para cumplir con lo requerido en la corrección

De cara a cumplir con el requisito del producto miniomo viable verdaderamente minimo he decidido simplificar mi codigo y centrarlo en solamente una funcion. Dicha función es la que extrae la fecha de una fila por ser esta la mas importante. La fecha funciona como un identificador. Esta funcion recorre una linea de codigo que se le pase y, usando una expresión regular, extrae el valor de la fecha. Se aportan también los test requeridos para esta. [Es un fix, nada nuevo añadido]

jorgelopez-ugr commented:
@gabrielherreraloz  @JJ listo, lo he simplificado. Ahora nos centramos solamente en la extracción del dato fecha. Uso una expresión regular y se aportan los test para demostrar su validez.

gabrielherreraloz reviewed (approved):
Lo veo mucho mejor así, corto, directo al grano de un PMV, y todo muy bien comentado y claro, tests variados también, pocos aunque tampoco da mucho juego al ser solo la fecha.
Mi aprobado lo tienes.

JJ reviewed (commented):
Por favor, revisa un poco.

jorgelopez-ugr committed:
Desarrollados y completados mis razonamientos en tests.md

Pertenece a la #40 y en ultima instancia a la #31. Se trata de un fix para ajustarse al feedback de la correciíon.

jorgelopez-ugr committed:
Continuando con el fix de la documentación de test

menciono el nombre de la biblioteca que uso y añado un link a la documentación oficial

jorgelopez-ugr committed:
fix porque habia un link repetido

ademas del link he movido la mencion a la biblioteca porque sentia que no tenia sentido mencionarla antes de la eleccion final

jorgelopez-ugr committed:
eliminación del codigo comentado

es un fix porque olvidé eliminar código que tenía comentado definiendo unos constructores que no son necesarios en el dominio de la solución. Era código usado en el  debug que olvidé eliminar

jorgelopez-ugr committed:
fix de correcciones, eliminacion de las funciones de error

eliminado Errores.js como se recomienda en la corrección. Visto ahora no sirve para nada. Como no aporta valor tenerlo en funciones aparte no tiene sentido añadir una capa de ruido intermedio.

jorgelopez-ugr committed:
fix eliminando comentarios

eliminados unos comentarios que no tenian sentido porque el propio codigo se explica por si solo

jorgelopez-ugr committed:
fix elimando la variable de entrada en html

La mera existencia de este bloque de html en codigo plano en el scraper es una mala práctica en si misma. Primeramente no es usado por ninguna funcion y segundo representa una variable que debería ser leida de el propio archivo. Tal paso intermedio añade ruido

jorgelopez-ugr committed:
fix unused imports

eliminados un par de imports para aserciones que no se usan

jorgelopez-ugr commented:
@JJ Listos los cambios a partir del feedback.

He desarrollado mucho más los razonamientos que me llevan a usar las herramientas que uso y eliminadas las definiciones.

He limpiado el archivo de Errores porque la verdad que no tenia mucho sentido, ahora disparo los errores directamente donde se los necesita.

He limpiado la variable del bloque html porque ni siquiera se usaba y no es una buena practica añadir ese paso intermedio si tenemos el archivo como tal.

He limpiado comentarios e imports que ya no se usaban despues de la simplificación masiva. Esto incuye también a los constructores debug que no usaba y estaban comentados.

La función sigue pasando los test sin problema.

He revisado que no haya código en el push que ya no exista. No se que ha pasado antes pero ya tenemos todo lo que necesitamos nada mas y nada menos.

Espero atentamente el feedback :+1:

JJ commented on docs/tests.md:
```diff
@@ -0,0 +1,78 @@
+    - Assertion Styles de Chai: [link](https://www.chaijs.com/guide/styles/)
+    - Necesitas extensiones e instalar paquetes nuevos con npm para poder usar la API assert. Definitivamente no es una integración nativa con Deno.
+
+- **Jest**
```
De veras, no le deis a la máquina de generar bibliotecas de aserciones. Jest **no es una biblioteca de aserciones**

JJ reviewed (changes requested):
Por favor, haced un pequeño esfuerzo por justificar bien las decisiones y buscar *vosotros mismos* las bibliotecas involucradas en la decisión. Si no lo hacéis, señalaré el error y pararé de corregir.

JJ commented:
Es más, deno tiene al menos dos bibliotecas de aserciones. La elección de herramientas bien documentada e investigada es una habilidad del ingeniero de software tan válida o más que el crear issues adecuados. Os rogaría que le diérais la importancia que merece.

jorgelopez-ugr committed:
fix al archivo de justificacion

se eliminan unos detalles y se añaden algunos enlaces para aportar mas informacion

jorgelopez-ugr committed:
se elimina informacion del README que ya no es valida

quedaba informacion que ha sido eliminada y se ha quitado de aqui tambien

jorgelopez-ugr committed:
fix por utilizacion de mayusculas

por lo comentado no es buena practica tener mayusculas en el nombre de los archivos, en este commit se produce un renombre de Precios.js a precios.js y Scraper.js a scraper.js

jorgelopez-ugr committed:
Closes #51 Se comprueba la validez de las fechas

Destacar que ahora en el scraper se extraen dias mes y año para hacer una comprobacion aparte. Con esta comprobacion nos aseguramos de que son fechas y en efecto fechas validas limitando el valor que pueden tomar los dias, los meses y los años a valores dentro de rangos razonables.

jorgelopez-ugr committed:
fix a un comentario sin valor

jorgelopez-ugr committed:
fix a las explicaciones por lo aclarado en clase

modificación al criterio para ser mas precisos en lo que determina que herramientas usar. Ahora ganan peso herramientas auditadas por el equipo de deno y se mantiene que no se necesiten archivos de configuración extra. Referente a la #40

jorgelopez-ugr commented:
@JJ He aplicado todos los fixes correspondientes a lo comentado en clase. Esta listo para revisión.

JJ commented on docs/tests.md:
```diff
@@ -0,0 +1,66 @@
+    - Assertion Styles de Chai: [link](https://www.chaijs.com/guide/styles/)
+    - Necesitas extensiones e instalar paquetes nuevos con npm para poder usar la API assert. Definitivamente no es una integración nativa con Deno.
+
+- **Paquete std/assert de jsr auditado por deno**
```
Es `@std/asssert`, pero bueno. `@std` es como se indican "espacios de nombres" en el ecosistema JS

JJ commented on lib/domain/scraper.js:
```diff
@@ -0,0 +1,43 @@
+        throw new Error("[Error] No se encontró una fecha válida en " + fila + ", error de formato.");
+    }
+
+    const dia = match[1];
```
Agh, ¿no has mirado la asignación por desestructuración?

JJ reviewed (approved):
👍 
De veras que tendríais que tener un poquito más de atención al detalle. Sois de hacer las cosas y listo. Hay que tratar de entender un poquito mejor el lenguaje y lo que lo hace especial o diferente de otros.
Además, conocer bien las construcciones idiomáticas te da siempre un edge por encima de otros en entrevistas. Que no vas a estar toda la vida donde estás.

jorgelopez-ugr committed:
fix usando asignacion por desestructuración de ES6

Para finalizar con la corrección del código se simplifican 3 lineas de codigo en 1 usando la asignacion por desestructuracion que permite asignar de forma multiple los valores de un array a una serie de variables como se especifica en la corrección.

jorgelopez-ugr committed:
fix añadiendo la @ a @std

continuando con la correccion pre-merge se añade la @ a @std/assert para ser correctos
